plugins/module_utils/foreman_helper.py

```diff
@@ -643,10 +644,238 @@ def exit_json(self, changed=False, **kwargs):
+        * entity_resolve (boolean): Set it to False to disable base entity resolution.
+        * entity_opts (dict): Dict of options for base entity. Same options can be provided for subentities described in entity_spec.
+    """
+    def __init__(self, argument_spec=None, **kwargs):
+        self.original_entity_spec = kwargs.pop('entity_spec', {})
+        self.sub_entities = {key: value for key, value in self.original_entity_spec.items()
+                             if value.get('resolve', True) and 'type' in value and value['type'] in ['entity', 'entity_list']}
+        self.entity_scope = kwargs.pop('entity_scope', None)
+        self.entity_key = kwargs.pop('entity_key', 'name')
+        self.entity_resolve = kwargs.pop('entity_resolve', True)
+        self.entity_opts = kwargs.pop('entity_opts', {})
+        if 'thin' not in self.entity_opts:
+            self.entity_opts['thin'] = False
+        if 'failsafe' not in self.entity_opts:
+            self.entity_opts['failsafe'] = True
+        if 'parent' in self.sub_entities:
+            self.entity_opts['search_by'] = 'title'
+        elif 'search_by' not in self.entity_opts:
+            self.entity_opts['search_by'] = self.entity_key
+
+        # Convert current class name from CamelCase to snake_case
+        class_name = re.sub(r'(?<=[a-z])[A-Z]|[A-Z](?=[^A-Z])', r'_\g<0>', self.__class__.__name__).lower().strip('_')
+        # Get entity name from snake case class name
+        self.entity_name = '_'.join(class_name.split('_')[1:-1])
+        p = apypie.Inflector()
+        self._entity_resource_path = p.pluralize(self.entity_name)
+        self.search = ""
+        super(ResolverEntityMixin, self).__init__(argument_spec=argument_spec, **kwargs)
+
+    def run(self, entity_dict=None, entity=None, connect=True, exit=True):
+        """ Connect, get entity_dict, resolve entities, ensure entity, remove sensitive data, manage parameters and exit.
+            You can provide custom entity_dict and entity if custom workflow is needed before doing most of listed actions.
+            If your module have some custom workflow, you can opt-out some actions:
+            * Connect by setting connect=False
+            * Exit by setting exit=False. In this case it will return the current entity
+        """
+        if connect:
+            self.connect()
+        if not entity_dict:
+            entity_dict = self.clean_params()
+
+        entity, entity_dict = self.resolve_entities(entity_dict, entity)
+
+        entity = self.ensure_entity(self._entity_resource_path, entity_dict, entity)
+
+        entity = self.remove_sensitive_fields(entity)
+        self.ensure_entity_parameters(entity, entity_dict)
+
+        if exit:
+            self.exit_json()
+        else:
+            return entity
+
+    def remove_sensitive_fields(self, entity):
+        """ Set fields with 'no_log' option to None """
+        if entity:
+            for blaclisted_field in self.blacklisted_fields:
```

Comment:
Typo (that's at least consistent with the next line):
```suggestion
            for blacklisted_field in self.blacklisted_fields:
```